### ISSUE TITLE
Inconsistent behaviour of zip comprehensions

### ISSUE THREAD

**Issue by AzamatB** [JuliaLang/julia#33168]
Consider
```Julia
julia> [rand() for _ in 1:3]
3-element Array{Float64,1}:
 0.084950477529397 
 0.5901110231368294
 0.5850548894960674

julia> [rand() for _ in zip(1:3, 1:5)] # expected this to work as the previous one
ERROR: DimensionMismatch("dimensions must match")
Stacktrace:
 [1] promote_shape at ./indices.jl:154 [inlined]
 [2] _promote_shape at ./iterators.jl:317 [inlined]
 [3] axes at ./iterators.jl:316 [inlined]
 [4] _array_for at ./array.jl:598 [inlined]
 [5] collect(::Base.Generator{Base.Iterators.Zip{Tuple{UnitRange{Int64},UnitRange{Int64}}},getfield(Main, Symbol("##57#58"))}) at ./array.jl:611
```
This failure does not make sense to me. Also if I turn the comprehension into loop it works again
```Julia
for _ in zip(1:3, 1:5)
   rand()
end
```

**Comment by TheBB** [comment]
Presumably fixed by https://github.com/JuliaLang/julia/pull/29927

**Comment by mschauer** [comment]
Fixed by #29927
